src/charm.py

```diff
@@ -167,19 +292,40 @@ def _generate_spring_boot_layer(self) -> dict:
             },
         }
 
-    def _service_reconciliation(self) -> None:
-        """Run the reconciliation process for pebble services.
+    def _get_sprint_boot_container_memory_constraint(self) -> typing.Optional[int]:
+        """Get the spring-boot-app container memory limit.
+
+        Return:
+            The memory limit of the spring-boot-app container in number of bytes. ``None`` if
+            there's no limit.
 
         Raises:
-            ReconciliationError: unrecoverable errors happen during the Java application type
-                detection process, requiring the main reconciliation process to terminate the
-                reconciliation early.
+            RuntimeError: Container spring-boot-app does not exist, it shouldn't happen since
+                this function checks the existence of the spring-boot-app first.
         """
+        # ensure that the Spring Boot container is up
+        self._spring_boot_container()
+        kubernetes.config.load_incluster_config()
+        client = kubernetes.client.CoreV1Api()
+        spec: kubernetes.client.V1PodSpec = client.read_namespaced_pod(
+            name=f"{self.app.name}-0", namespace=self.model.name
+        ).spec
+        for container in spec.containers:
+            if container.name != "spring-boot-app":
```

Comment:
what about using `next` and an iteration over the containers with a filter handling the case where no container is found?

Reply:
Using `next` dose not simplify the statements here. For `next`, a single use filter function has to be defined, and the `StopIteration` exception has to be handled. That makes the `next` method as or more complex than for-loop.
Also, I think this is against the zen of Python.

Reply:
It is also possible to pass an iterator to the `next` function, e.g., `next((container for container in spec.containers if container.name == "spring-boot-app"), None), then you get the first container that matches the condition or None if no container matches the condition, so no need to handle `StopIteration`, no need to write single use filter function. What do you think?

Reply:
This looks better. Updated, thanks!

Reply:
For using `all` in integration tests to replace for loop, pytest seems to not show the comparison between expect and actual value at test failure when using `all`.

Reply:
Yes that is a drawback, in the case of assert in tests it is probably better to use the for loop

Reply:
Or you can add a nice error message like `assert any(<expression>), <message>`

Reply:
The message approach is that I don't know which item in the iterator will fail, so I have to include all items in the message. That can result in a very long message.